plugins/woocommerce/src/Blocks/Patterns/PTKPatternsStore.php

```diff
@@ -32,42 +33,88 @@ public function __construct( PTKClient $ptk_client ) {
+			add_action( 'deactivated_plugin', array( $this, 'flush_cached_patterns' ), 10, 2 );
+			add_action( 'upgrader_process_complete', array( $this, 'fetch_patterns_on_plugin_update' ), 10, 2 );
+
+			// This is the scheduled action that takes care of flushing and re-fetching the patterns from the PTK API.
+			add_action( 'fetch_patterns', array( $this, 'fetch_patterns' ) );
 		}
 	}
 
 	/**
-	 * Register block patterns from the Patterns Toolkit.
+	 * Resets the cached patterns when the `woocommerce_allow_tracking` option is disabled.
+	 * Resets and fetch the patterns from the PTK when it is enabled (if the scheduler
+	 * is initialized, it's done asynchronously via a scheduled action).
 	 *
 	 * @return void
 	 */
-	public function get_patterns() {
-		$patterns = get_transient( self::TRANSIENT_NAME );
+	public function flush_or_fetch_patterns() {
+		if ( $this->allowed_tracking_is_enabled() ) {
+			$this->schedule_fetch_patterns();
+			return;
+		}
 
-		// Only if the transient is not set, we fetch the patterns from the PTK.
-		if ( false === $patterns ) {
-			$patterns = $this->ptk_client->fetch_patterns(
-				array(
-					'categories' => array( 'intro', 'about', 'services', 'testimonials' ),
-				)
+		$this->flush_cached_patterns();
+	}
+
+	/**
+	 * Schedule an async action to fetch the PTK patterns when the scheduler is initialized.
+	 *
+	 * @return void
+	 */
+	private function schedule_fetch_patterns() {
+		if ( did_action( 'action_scheduler_init' ) ) {
+			$this->schedule_action_if_not_pending( 'fetch_patterns' );
+		} else {
+			add_action(
+				'action_scheduler_init',
+				function () {
+					$this->schedule_action_if_not_pending( 'fetch_patterns' );
+				}
 			);
+		}
+	}
 
-			if ( is_wp_error( $patterns ) ) {
-				wc_get_logger()->warning(
-					sprintf(
-					// translators: %s is a generated error message.
-						__( 'Failed to get the patterns from the PTK: "%s"', 'woocommerce' ),
-						$patterns->get_error_message()
-					),
-				);
-				return;
-			}
+	/**
+	 * Schedule an action if it's not already pending.
+	 *
+	 * @param string $action The action name to schedule.
+	 * @return void
+	 */
+	private function schedule_action_if_not_pending( $action ) {
+		if ( as_has_scheduled_action( $action ) ) {
+			return;
+		}
 
-			$patterns = $this->filter_patterns( $patterns, self::EXCLUDED_PATTERNS );
+		as_schedule_single_action( time(), $action );
+	}
 
-			set_transient( self::TRANSIENT_NAME, $patterns );
+	/**
+	 * Get the patterns from the Patterns Toolkit cache.
+	 *
+	 * @return array|WP_Error
+	 */
+	public function get_patterns() {
+		$patterns = get_transient( self::TRANSIENT_NAME );
+
+		// Only if the transient is not set, we schedule fetching the patterns from the PTK.
+		if ( false === $patterns ) {
+			$this->schedule_fetch_patterns();
+			return new WP_Error(
+				'patterns_store_error',
+				__( 'Failed to get the cached PTK patterns.', 'woocommerce' )
+			);
```

Comment:
I'm not sure that we should return an error here 🤔 , we could just return `undefined` (maybe I missed something)

Reply:
Fixed it here https://github.com/woocommerce/woocommerce/pull/47551/commits/cd1ce72f4ec5cc64b2c1b5aadce877ed40bbdae9